src/TypeSystem/Scope/TernaryBranchScope/TernaryBranchScope.php

```diff
@@ -0,0 +1,66 @@
+<?php
+
+/**
+ * PackageFactory.ComponentEngine - Universal View Components for PHP
+ *   Copyright (C) 2022 Contributors of PackageFactory.ComponentEngine
+ *
+ *   This program is free software: you can redistribute it and/or modify
+ *   it under the terms of the GNU General Public License as published by
+ *   the Free Software Foundation, either version 3 of the License, or
+ *   (at your option) any later version.
+ *
+ *   This program is distributed in the hope that it will be useful,
+ *   but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *   GNU General Public License for more details.
+ *
+ *   You should have received a copy of the GNU General Public License
+ *   along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+declare(strict_types=1);
+
+namespace PackageFactory\ComponentEngine\TypeSystem\Scope\TernaryBranchScope;
+
+use PackageFactory\ComponentEngine\Parser\Ast\ExpressionNode;
+use PackageFactory\ComponentEngine\Parser\Ast\TypeReferenceNode;
+use PackageFactory\ComponentEngine\TypeSystem\Inferrer\InferredTypes;
+use PackageFactory\ComponentEngine\TypeSystem\Inferrer\TypeInferrer;
+use PackageFactory\ComponentEngine\TypeSystem\Inferrer\TypeInferrerContext;
+use PackageFactory\ComponentEngine\TypeSystem\ScopeInterface;
+use PackageFactory\ComponentEngine\TypeSystem\TypeInterface;
+
+final class TernaryBranchScope implements ScopeInterface
+{
+    private function __construct(
+        private readonly InferredTypes $inferredTypes,
+        private readonly ScopeInterface $parentScope
+    ) {
+    }
+
+    public static function forTruthyBranch(ExpressionNode $conditionNode, ScopeInterface $parentScope): self
+    {
+        return new self(
+            (new TypeInferrer($parentScope))->inferTypesInCondition($conditionNode, TypeInferrerContext::TRUTHY),
+            $parentScope
+        );
+    }
+
+    public static function forFalsyBranch(ExpressionNode $conditionNode, ScopeInterface $parentScope): self
+    {
+        return new self(
+            (new TypeInferrer($parentScope))->inferTypesInCondition($conditionNode, TypeInferrerContext::FALSY),
+            $parentScope
+        );
+    }
```

Comment:
As mentioned in the comments for `TypeInferrer.php`, the `TernaryOperationTypeResolver` class should take care of instantiating one `TypeNarrower` that both `TernaryBranchScope`s can share.

This has consequences for the signature of these two methods and...

Reply:
nvm. (see: https://github.com/PackageFactory/PackageFactory.ComponentEngine/pull/14#discussion_r1179451334)